MetroidPrimeInterface.py

```diff
@@ -484,3 +591,35 @@ def sync_artifact_layers(self) -> bool:
                         self.set_layer_active(
                             ARTIFACT_TEMPLE_ROOM_INDEX, layer_id, item.current_amount > 0)
                         changed = True
+
+    def reset_relay_tracker_cache(self):
+        self.relay_trackers = None
+
+    def update_relay_tracker_cache(self):
+        if self.relay_trackers is None:
+            self.relay_trackers = {}
+            world_state_array = struct.unpack(">I", self.dolphin_client.read_pointer(GAMES[self.current_game]["game_state_pointer"], 0x94, struct.calcsize(">I")))[0]
+            world_states = [world_state_array + i * 0x18 for i in range(7)]
+            for world_state in world_states:
+                mlvl = struct.unpack(">I", self.dolphin_client.read_address(world_state, struct.calcsize(">I")))[0]
+                self.relay_trackers[f'{mlvl:X}'] = {
+                    "address": struct.unpack(">I", self.dolphin_client.read_pointer(world_state + 8, 0, struct.calcsize(">I")))[0],
+                    "count": 0,
+                    "memory_relays": [],
+                }
+        for _, relay_tracker in self.relay_trackers.items():
+            relay_tracker["count"] = struct.unpack(">I", self.dolphin_client.read_address(relay_tracker["address"], struct.calcsize(">I")))[0]
+            relay_tracker["memory_relays"] = struct.unpack(">"+("I" * relay_tracker["count"]), self.dolphin_client.read_address(relay_tracker["address"] + 4, relay_tracker["count"] * struct.calcsize(">I")))

```

Comment:
I wouldn't mind having the 4 here and the 8 on 606 also named if you're alright with that. (I'm able to follow most of this but get a little lost in what the offsets etc are doing 😅)

Reply:
The one one 612 as well

Reply:
That unpack counter is pretty cool 🔥